blaze/expr/core.py

```diff
@@ -411,21 +415,72 @@ def path(a, b):
     yield a
 
 
+def ordered_intersect(a, b):
+    """Set intersection of two sequences that preserves order.
+
+    Parameters
+    ----------
+    a, b : Sequence
+
+    Returns
+    -------
+    list
+
+    Examples
+    --------
+    >>> ordered_intersect('abcd', 'cdef')
+    ['c', 'd']
+    >>> ordered_intersect('bcda', 'bdfga')
+    ['b', 'd', 'a']
+    >>> ordered_intersect('zega', 'age')  # first sequence determines ordering
+    ['e', 'g', 'a']
+    """
+    common = set(a) & set(b)
+    return [x for x in unique(concat((a, b))) if x in common]
+
+
 def common_subexpression(*exprs):
     """ Common sub expression between subexpressions
 
     Examples
     --------
 
-    >>> from blaze.expr import symbol, common_subexpression
-
+    >>> from blaze.expr import symbol
     >>> t = symbol('t', 'var * {x: int, y: int}')
     >>> common_subexpression(t.x, t.y)
     t
     """
-    sets = [set(subterms(t)) for t in exprs]
-    return builtins.max(set.intersection(*sets),
-                        key=compose(len, str))
+    # only one expression has itself as a common subexpression
+    if len(exprs) == 1:
+        return exprs[0]
+
+    # get leaves for every expression
+    all_leaves = [expr._leaves() for expr in exprs]
+
+    # leaves common to all expressions
+    leaves = set.intersection(*map(set, all_leaves))
```

Comment:
this fails if there are no exprs, one way to fix this is to define the function like:

```python
def common_subexpression(expr, *exprs):
    exprs = (expr,) + exprs
```
This will make it raise the correct type error if the one required argument is not passed.